### ISSUE TITLE
Support for rMBP 15" 2012 (Nvidia GT650m) / macOS 10.12.3

### ISSUE THREAD

**Issue by austrianhack** [Floris497/mac-pixel-clock-patch-V2#144]
Hi Floris! 

Please add support for my Nvidia GT 650m, thanks in advance!

Last login: Sat Feb 25 22:43:31 on ttys000
Sebastians-MacBook-Pro:~ sebastianvogrin$ /Users/sebastianvogrin/Desktop/NVIDIA-patcher.command 
SIP looks to be disabled, all good!
Keep SIP disabled, if enabled again SIP prevent the Nvidia Driver from loading

Unknown version of the Nvidia driver found..
---- BEGINNING MD5 HASH SUMS ---- version: 10.12.3 build:16D32

     NVDAGK100Hal: 53b038b824ae3941e7df1f30e41250dc
 NVDAGK100Hal.bak: 53b038b824ae3941e7df1f30e41250dc

---- ENDING MD5 HASH SUMS -------

**Comment by Floris497** [comment]
Please force with newest version :)

**Comment by austrianhack** [comment]
Thanks for the quick response! So I unpatched (as recommended to keep original backup) and then patched v2:

Sebastians-MacBook-Pro:~ sebastianvogrin$ /Users/sebastianvogrin/Desktop/NVIDIA-patcher.command patch v2
SIP looks to be disabled, all good!
Keep SIP disabled, if enabled again SIP prevent the Nvidia Driver from loading
Patching NVDAGK100Hal with patch version 2
The code is still the same.. Patch did seem to run, but was probably from the wrong version..
If you are running an new os run /Users/sebastianvogrin/Desktop/NVIDIA-patcher.command md5 and ask the maintainer of this script to add support for your system

...

Still stuck on 30 Hz :-(

**Comment by Floris497** [comment]
Could you try to run it with v3?

**Comment by austrianhack** [comment]
```
Sebastians-MacBook-Pro:~ sebastianvogrin$ /Users/sebastianvogrin/Desktop/NVIDIA-patcher.command patch v3
SIP looks to be disabled, all good!
Keep SIP disabled, if enabled again SIP prevent the Nvidia Driver from loading
Patching NVDAGK100Hal with patch version 3
The code of the driver has changed, the patch was probably successful

```

Still just 30 Hz ... :-(

**Comment by Floris497** [comment]
And if you also patch IOKit?

**Comment by Floris497** [comment]
Btw, what cable are you using?

**Comment by austrianhack** [comment]
I am using a mDP to DP cable but I've also tried a mDP to HDMI adaptor then using a HDMI 2.0 cable ...

Both configs give me 30 Hz only.

BUT I'll try the IOKit patch like you said...although I thought it wasn't necessary for 10.12 (?)

**Comment by austrianhack** [comment]
Still no success :-( 

tried IOKit v2

**Comment by austrianhack** [comment]
Anything else that comes to your mind? 

Would be much appreciated :-)

**Comment by Floris497** [comment]
I'm thinking 💭
I will let you know

**Comment by austrianhack** [comment]
Thanks for the effort Floris.. :-)

**Comment by michal-olszewski** [comment]
rMBP 15 from 2012 (GT650m) has Thunderbolt 1, which encapsulates DP 1.1a - bandwidth is too small to obtain 4k@60Hz.

**Comment by Floris497** [comment]
My macmini 2011 does 4k@40Hz but no 4k at all without patch..

**Comment by michal-olszewski** [comment]
When I used such a 2012 rMBP it run 4k@30Hz on both HDMI and DP out of box, but the highest refresh rate attainable with some tweaking was around 40Hz (over DP).

**Comment by austrianhack** [comment]
Right, didn't think about that michal! Well, no 4k @ 60 Hz for me ;-P

Luckily I still got a hackintosh too which is connected to the 4k monitor right now :-)

**Comment by jeroenson** [comment]
So this will be the same for my early 2013 MPB 15" ?

**Comment by Floris497** [comment]
Doesn't 2013 have 4k by default? It think 2013 might be able to do 60Hz

**Comment by jeroenson** [comment]
Late 2013 has 4K by default. I tried to patch mine yesterday with the CoreDisplay patch. The patch succeeded but only managed to get it to run 4K 30Hz (mDP to DP cable).

**Comment by Floris497** [comment]
Using switchresx?

**Comment by jeroenson** [comment]
Added the custom resolution 4K@60Hz but I can't select it. 
<img width="1243" alt="schermafbeelding 2017-02-28 om 17 06 44" src="https://cloud.githubusercontent.com/assets/10088007/23453966/c7ab9820-fe6a-11e6-8db0-9a40ef0e2943.png">
<img width="792" alt="schermafbeelding 2017-02-28 om 17 02 43" src="https://cloud.githubusercontent.com/assets/10088007/23453968/c9983b52-fe6a-11e6-8bef-bf840c011693.png">

**Comment by Floris497** [comment]
Wat staat er op de 'custom resolutions' pagina? Staat de resolutie daar op 'active'?

**Comment by jeroenson** [comment]
Check het vanavond even zit nu niet achter dit scherm. Het was niet mogelijk om de resolutie te activeren ook niet na een reboot. De correcte melding moet ik je nog even schuldig blijven.

**Comment by Floris497** [comment]
Ah nee, als er 'inactive - unsupported' staat dan werkt het niet..
